pkg/v1/partial/index.go

```diff
@@ -0,0 +1,84 @@
+// Copyright 2020 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package partial
+
+import (
+	"bytes"
+	"errors"
+
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+)
+
+// WithRawIndexManifest defines the subset of v1.ImageIndex used by ImageByArch.
+type WithRawIndexManifest interface {
+	RawManifest() ([]byte, error)
+	Image(v1.Hash) (v1.Image, error)
+}
+
+// IndexManifest is a helper for implementing v1.ImageIndex.
+func IndexManifest(w WithRawIndexManifest) (*v1.IndexManifest, error) {
+	b, err := w.RawManifest()
+	if err != nil {
+		return nil, err
+	}
+	return v1.ParseIndexManifest(bytes.NewReader(b))
+}
+
+// ImageByPlatform returns the v1.Image belonging to the v1.ImageIndex with the
+// matching Platform specification.
+func ImageByPlatform(w WithRawIndexManifest, p v1.Platform) (v1.Image, error) {
+	im, err := IndexManifest(w)
+	if err != nil {
+		return nil, err
+	}
+	for _, manifest := range im.Manifests {
+		if manifest.Platform != nil && platformsEqual(*manifest.Platform, p) {
+			return w.Image(manifest.Digest)
+		}
+	}
+	return nil, errors.New("failed to find image for platform in index")
+}
+
+func platformsEqual(a, b v1.Platform) bool {
```

Comment:
This is too naive -- see the equivalent logic remote.

We basically need to answer the question: is there an image that satisfies the requirements of the given platform. E.g. if there is an image with the superset of features, that's fine.

Reply:
Oh yeah, absolutely right. Now just wholesale stealing the logic from pkg/v1/remote/ 😄